### PR TITLE
[WIP] Use SpanID instead of Partition, Offset

### LINKED ISSUE #483: contrib/Shopify/sarama: bug in AsyncProducer
bugs are here.
- https://github.com/DataDog/dd-trace-go/blob/v1/contrib/Shopify/sarama/sarama.go#L206 

According to the document of `sarama.ProducerMessage`, `msg.Offset` and `msg.Partition` is not valid before the message was successfully delivered.
- https://godoc.org/github.com/Shopify/sarama#ProducerMessage

for this reason, `finishProducerSpan` can't be called successfully.

### PR DESCRIPTION
**Pull request by HatsuneMiku3939**
* only successfully transmitted message have valid Partition, Offset

Closes #483

### CODE REVIEW

File: contrib/Shopify/sarama/sarama.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ type syncProducer struct {
 func (p *syncProducer) SendMessage(msg *sarama.ProducerMessage) (partition int32, offset int64, err error) {
 	span := startProducerSpan(p.cfg, p.version, msg)
 	partition, offset, err = p.SyncProducer.SendMessage(msg)
-	finishProducerSpan(span, partition, offset, err)
+	finishProducerSpan(span, msg, err)
```
Comment (HatsuneMiku3939) [inline contrib/Shopify/sarama/sarama.go]: I'm not sure sarama's SyncProducer modify original message with the valid Partition, Offset .

File: contrib/Shopify/sarama/sarama.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func (p *syncProducer) SendMessages(msgs []*sarama.ProducerMessage) error {
 	}
 	err := p.SyncProducer.SendMessages(msgs)
 	for i, span := range spans {
-		finishProducerSpan(span, msgs[i].Partition, msgs[i].Offset, err)
+		finishProducerSpan(span, msgs[i], err)
```
Comment (HatsuneMiku3939) [inline contrib/Shopify/sarama/sarama.go]: I'm not sure sarama's SyncProducer modify original message with the valid Partition, Offset .

File: contrib/Shopify/sarama/sarama.go [hunk 3]
```diff
@@ -192,48 +192,47 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
...
 			case err, ok := <-p.Errors():
 				if !ok {
 					// producer was closed
 					return
 				}
-				key := spanKey{err.Msg.Topic, err.Msg.Partition, err.Msg.Offset}
-				if span, ok := spans[key]; ok {
-					delete(spans, key)
-					finishProducerSpan(span, err.Msg.Partition, err.Msg.Offset, err.Err)
+				if spanctx, spanFound := getSpanContext(msg); spanFound {
```
Comment (HatsuneMiku3939) [inline contrib/Shopify/sarama/sarama.go]: I will fix it later

**Comment by cgilmour** [comment]
Could #483 be fixed by changing [this line](https://github.com/DataDog/dd-trace-go/blob/v1/contrib/Shopify/sarama/sarama.go#L215) to `span.Finish()` ?

**Comment by HatsuneMiku3939** [comment]
> Could #483 be fixed by changing [this line](https://github.com/DataDog/dd-trace-go/blob/v1/contrib/Shopify/sarama/sarama.go#L215) to `span.Finish()` ?

#483 will be fixed by changing [this line](https://github.com/DataDog/dd-trace-go/blob/v1/contrib/Shopify/sarama/sarama.go#L210)

**Comment by gbbr** [comment]
@HatsuneMiku3939 can we close this?

**Comment by adw1n** [comment]
For what it's worth I tried using `saramatrace.WrapAsyncProducer` with no success - no traces showed up in Datadog. I patched `gopkg.in/DataDog/dd-trace-go.v1/contrib/Shopify/sarama` to include these changes and traces started flowing in.

**Comment by knusbaum** [comment]
I am closing this as #738 is likely to go in soon.

**Pull request closed without merging**